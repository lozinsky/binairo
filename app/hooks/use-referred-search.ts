import { type Path, createPath } from '@remix-run/react';
import { useMemo } from 'react';

export function useReferredSearch(search: string, referrerPath: Partial<Path> | null) {
  const referredSearch = useMemo(() => {
    const searchParams = new URLSearchParams(search);

    if (referrerPath !== null) {
      searchParams.set('referrer', createPath(referrerPath));
    }

    return searchParams.toString();
  }, [referrerPath, search]);

  return referredSearch;
}
