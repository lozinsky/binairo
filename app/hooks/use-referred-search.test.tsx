/* @vitest-environment jsdom */

import type { ReactNode } from 'react';

import { renderHook } from '@testing-library/react';
import { type Location, type Path, useLocation } from 'react-router';
import { createRoutesStub } from 'react-router';
import { expect, test } from 'vitest';

import { useReferredSearch } from './use-referred-search';

test.each<{ location: Partial<Location>; referrerPath: null | Partial<Path> }>([
  {
    location: { search: '' },
    referrerPath: null,
  },
  {
    location: { search: '' },
    referrerPath: { pathname: '/', search: '' },
  },
  {
    location: { search: '' },
    referrerPath: { pathname: '/game', search: '?referrer=%2F' },
  },
  {
    location: { search: '?analyze' },
    referrerPath: null,
  },
  {
    location: { search: '?analyze' },
    referrerPath: { pathname: '/game', search: '?referrer=%2F' },
  },
])('returns referred search', ({ location, referrerPath }) => {
  function Wrapper({ children }: { children: ReactNode }) {
    const RouterStub = createRoutesStub([{ Component, path: '/' }]);

    function Component() {
      return children;
    }

    return <RouterStub initialEntries={[location]} />;
  }

  const { result } = renderHook(
    () => {
      const location = useLocation();
      const referredSearch = useReferredSearch(location.search, referrerPath);

      return referredSearch;
    },
    { wrapper: Wrapper },
  );

  expect(result.current).toMatchSnapshot();
});
