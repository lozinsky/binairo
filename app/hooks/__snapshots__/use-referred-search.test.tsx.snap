// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns referred search 1`] = `""`;

exports[`returns referred search 2`] = `"?referrer=%2F"`;

exports[`returns referred search 3`] = `"?referrer=%2Fgame%3Freferrer%3D%252F"`;

exports[`returns referred search 4`] = `"?analyze="`;

exports[`returns referred search 5`] = `"?analyze=&referrer=%2Fgame%3Freferrer%3D%252F"`;
